### PR TITLE
RUMM-555 Auto-tracing for Objective-C added

### PR DESCRIPTION
**Pull request by buranmert**
### What and why?

`DatadogObjc` needs to provide auto-instrumentation for Tracing as `Datadog` already does

### How?

We provide `set(tracedHosts: Set<String>)` in `DDConfigurationBuilder` interface.
This method sets `tracedHosts` in `Datadog.Configuration` which leads to automatic tracing in given hosts, just as `Datadog` does in Swift

#### Passing `nil` parameters to swizzled methods

Unlike in Swift, we can pass `nil` as `url`, `request` and/or `completionHandler` to swizzled methods in ObjC.
Although this never should happen, if that happens the app crashes with only `exc_bad_access`: our users won't even have a clue what went wrong!

We fixed that by changing the method signature in swizzle implementation
```swift
// before
(URLSession, Selector, URL, @escaping CompletionHandler)
// after
(URLSession, Selector, URL?, CompletionHandler?)
```
This change doesn't change the memory layout of those types at runtime, that's why it doesn't crash and works properly ✅ 
For more info and understand memory layout of `Enum` types, you can refer to [Swift source code](https://github.com/apple/swift/blob/master/include/swift/ABI/Enum.h#L48)

#### Executing swizzled `resume` more than once

⚠️ This edge case does NOT exist in any known scenarios, that's why I can't add tests for it ⚠️ 

```swift
class FirstSubclass: URLSessionTask { ... }
class SecondSubclass: FirstSubclass { ... }

// 1 - dataTask is called, FirstSubclass instance is returned and its resume is swizzled
let first = session.dataTask(with: someURL, ...) -> FirstSubclass
// 2 - dataTask is called again, SecondSubclass instance is returned and its resume is swizzled
let second = session.dataTask(with: anotherRequest, ...) -> SecondSubclass
// 3 - swizzled resume imp will be executed twice
second.resume() -> first.resume()
```

The fix is simpler than understanding the problem: now we consume payloads so that they can only be used once ✅ 
_(I noticed that case while looking at issues in `InterposeKit`, I can't find the related issue now, I'll put the link once I can)_

### Review checklist

- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [x] Make sure each commit and the PR mention the Issue number or JIRA reference

### CODE REVIEW

File: Datadog/Datadog.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -1428,6 +1438,7 @@
 					};
 					61133B8A242393DE00786299 = {
 						CreatedOnToolsVersion = 11.3.1;
+						LastSwiftMigration = 1150;
```
Comment (ncreated) [inline Datadog/Datadog.xcodeproj/project.pbxproj:1441]: Any idea where this comes from? Do you use `Xcode 11.5`?

File: Tests/DatadogTests/Datadog/Mocks/TracingFeatureMocks.swift [hunk 2]
```diff
@@ -213,8 +214,10 @@ class SpanOutputMock: SpanOutput {
     }
 
     var recorded: Recorded? = nil
+    let expectation = XCTestExpectation(description: "SpanRecorder expectation")
 
     func write(ddspan: DDSpan, finishTime: Date) {
         recorded = Recorded(span: ddspan, finishTime: finishTime)
+        expectation.fulfill()
     }
```
[the remark this answers is not in the document]
Comment (buranmert) [reply Tests/DatadogTests/Datadog/Mocks/TracingFeatureMocks.swift:222]: this lets us wait for the span to be recorded. we don't have to wait for this expectation.

use case:
we call the original `completion` first and then our `taskObserver` in `completionHandler`
```swift
let modifiedCompletion: CompletionHandler = { origData, origResponse, origError in
  impCompletion?(origData, origResponse, origError)
  taskObserver?(.completed(origResponse, origError))
}
```
the idea is executing the original code of host app as soon as possible.
we don't need that in 
1. `URLSessionSwizzlerTests` because we inject an `observationCompleted` expectation via our custom `Interceptor`
2. `TracingIntegrationTests` because we sleep the thread for a while

so, we can fix that by one of the following:
1. change the order of  `impCompletion` and `taskObserver` in swizzled implementation?
2. inject custom `Interceptor` instead of auto-instrument with `Datadog.initialize`?
3. add this expectation into `SpanMockRecorder`?

wdyt?

File: Sources/Datadog/Datadog.swift [hunk 3]
```diff
@@ -186,6 +186,8 @@ public class Datadog {
         // Then, deinitialize features:
         LoggingFeature.instance = nil
         TracingFeature.instance = nil
+        // Don't forget to TracingAutoInstrumentation.instance.swizzler.unswizzle()
+        // before calling deinitializeOrThrow() in tests!
```
[the remark this answers is not in the document]
Comment (buranmert) [reply Sources/Datadog/Datadog.swift:190]: we have `unswizzle` extensions in unit test target, we can't call it here in main target.
what do you mean by explicitly fixing?

File: Datadog/Datadog.xcodeproj/project.pbxproj (hunk 1 above)
Comment (buranmert) [reply Datadog/Datadog.xcodeproj/project.pbxproj:1441]: no i don't know how/why that's here 🤷‍♂️ 
yes, xcode 11.5

File: Tests/DatadogTests/DatadogObjc/NSURLSessionSwizzlingTests.swift [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        if let task = swizzledTask {
+            taskObservation = task.observe(\.state) { observedTask, _ in
+                if case URLSessionTask.State.completed = observedTask.state {
+                    expectation.fulfill()
+                }
+            }
+            task.resume()
+            wait(for: [expectation], timeout: timeout)
+        }
+    }
```
[the remark this answers is not in the document]
Comment (buranmert) [reply Tests/DatadogTests/DatadogObjc/NSURLSessionSwizzlingTests.swift:184]: in iOS 13.5 it returns `nil`
this code is removed ✅

File: Tests/DatadogTests/DatadogObjc/NSURLSessionSwizzlingTests.swift [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+
+        var originalException: NSError? = nil
+        var originalTask: URLSessionTask? = nil
+        do {
+            try objcExceptionHandler.rethrowToSwift {
+                originalTask = NSURLSessionBridge.session(self.session, dataTaskWith: nilRequest, completionHandler: nil)
+            }
+        } catch {
+            originalException = error as NSError
+        }
```
[the remark this answers is not in the document]
Comment (buranmert) [reply Tests/DatadogTests/DatadogObjc/NSURLSessionSwizzlingTests.swift]: the idea was passing if both are `nil` at the same time _"if original doesn't throw, swizzled should not throw neither"_
but in iOS 13.5 it always throws ✅ 
this code is changed 👍

File: Sources/Datadog/Datadog.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (ncreated) [reply Sources/Datadog/Datadog.swift:190]: I mean - I see no reason for this comment 🙂. It's a thing we only have to do in `TracingAutoInstrumentationTests`, no other places need it. It's misleading to put it there.

File: Tests/DatadogTests/Datadog/Mocks/TracingFeatureMocks.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (ncreated) [reply Tests/DatadogTests/Datadog/Mocks/TracingFeatureMocks.swift:222]: > 1. change the order of impCompletion and taskObserver in swizzled implementation?
> 2. inject custom Interceptor instead of auto-instrument with Datadog.initialize?
> 3. add this expectation into SpanMockRecorder?

I'd go with 2nd option, as instantiating the entire SDK (`DDDatadog.initialize()`) is not necessary for this unit test and produces many side effects (all the features are enabled, so their upload workers do the job of uploading data written by auto-instrumentation tracer). To avoid any sort of flakiness, unit tests should be instantiating / mocking only the necessary components - that's what `TracingFeature.mock()` and `LoggingFeature.mock()` are meant for.

And if we want to make `SpanOutputMock` wait on expectation, let's either make it configurable:
```swift
SpanOutputMock(expectation: ...)
```
or decouple it entirely from `SpanOutputMock`:
```swift
spanOutputMock.whenRecorded = { expectation.fulfill() }
```

Otherwise, the `SpanOutputMock` usage is unobvious, as it always holds an expectation which is hidden from its interface. If using it for "no span is sent" scenario, this will fail.

[the remark this answers is not in the document]
Comment (buranmert) [reply Tests/DatadogTests/Datadog/Mocks/TracingFeatureMocks.swift:222]: `whenRecorded` is a good idea 👍 
actually i should add only ObjC edge case tests, the rest is already tested in `TracingAutoInstrumentationTests` and `URLSessionSwizzlerTests` ✅ 

now i'm 
1. moving `_edgeCase`s to `URLSessionSwizzlerTests.swift`
2. reverting `SpanOutputMock`
3. removing this `NSURLSessionSwizzlingTests.swift`
makes sense?

File: Datadog/Datadog.xcodeproj/project.pbxproj (hunk 1 above)
Comment (buranmert) [reply Datadog/Datadog.xcodeproj/project.pbxproj:1438]: oh i think i know why that happened 💡 
i also have `xcode-beta` and the project was probably opened with it by default when i double-click on `xcworkspace`
because i just realized my `open by default` was changed to `xcode-beta` without any explicit action

removing this line ✅
